### PR TITLE
fixing transpose scheduler merge dimension issue

### PR DESCRIPTION
**Pull request by jjsjann123**
fixes: #675 (Although the problem only shows up after a few other PR that refactors transpose vectorization computation)

Transpose scheduler should merge dimensions in the expected order (from inner to outer on each group).
Otherwise generated kernel would violate alignment requirement for vectorization.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

File: csrc/scheduler/utils.cpp [hunk 1]
```diff
@@ -170,21 +170,36 @@ std::optional<size_t> mergeDims(
...
-    auto actual_right = offset-- + *right;
-    tv->merge((int)left, (int)actual_right);
-    for (auto& i : to_update) {
-      if (i == actual_right) {
-        i = left;
-      } else if (i > actual_right) {
-        i--;
+  auto inner = to_merge[0];
+
+  // NOTE: we need to merge dimensions in the order specified by `to_merge`,
```
Comment (naoyam) [inline csrc/scheduler/utils.cpp:175]: Can you please update the comment of this function and mention that merge is done in the order of `to_merge`?

File: csrc/scheduler/utils.cpp [hunk 2]
```diff
@@ -170,21 +170,36 @@ std::optional<size_t> mergeDims(
...
+  // otherwise this could results in misaligned memory access due to indexing.
+  // This is because we compute vectorization width before applying scheduling
+  // transformations.
+  for (size_t i = 1; i < to_merge.size(); i++) {
+    auto outer = to_merge[i];
+    if (outer > inner) {
+      tv->reorder({{inner, outer}, {outer, inner}});
+      std::swap(inner, outer);
+    }
```
Comment (naoyam) [inline csrc/scheduler/utils.cpp:184]: If a `to_merge` has `{1, 0}`, shouldn't we just do `merge(0, 1)`? Otherwise, it seems to contradict with the comment above?

**Review by zasdfgbnm — COMMENTED** [review]
Could you modify the `FusionMergeDims_CUDA` test to make sure this issue is covered?

File: csrc/scheduler/utils.cpp (hunk 2 above)
Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:184]: So dimensions in `to_merge` is ordered from inner to outer. When we have the next element in `to_merge` that is larger then `inner`, i.e.
```
tv[i0, i1]
outer = 1
inner = 0
```
We would want to have the merge result be tv[i1 * i0].

If we just do `tv->merge(0, 1)`, that will give us `tv[i0 * i1]`. Hence the reordering.

Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:184]: Let me put the comment in the code as well

Comment (naoyam) [reply csrc/scheduler/utils.cpp:184]: Maybe my confusion is, when you say:

> So dimensions in to_merge is ordered from inner to outer

Is the ordering of the inner and outer referring to the ordering in the domain? Or does it refer to the ordering in merge?

Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:184]: This is indeed very confusing :confused: 

When I talk about `inner to outer`, I was referring to the order in the `merged iter domain`.
so we want the end result to be something like
`tv->axis(to_merge[i-1]) * tv->axis(to_merge[i-2]) * ... * tv->axis(to_merge[0])`

^^^ this part is what inner vs outer.

In the mean time, when we merge `to_merge[k]` with `to_merge[k-1]`, there's no guarantee that the two item would be in any order. i.e. if to_merge[k] is larger than `to_merge[k-1]`, we need to reorder it, so we'll have the merged result be in `to_merge[k] * to_merge[k-1]` instead of the other way around.

Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:184]: actually, I think the last merged dimension doesn't seem to be at the right position... hmmm... interesting...

Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:184]: so this is still functional, but it's unsatisfying that the merged dimension would just be the outer most in the original leaf domain. `*std::min_element(to_merge.begin(), to_merge.end())` instead of the logical order.

Maybe I missed out how to do that properly in merge/reorder.

File: test/test_gpu_utils.cpp [hunk 3]
```diff
@@ -59,6 +59,11 @@ TEST_F(NVFuserTest, FusionMergeDims_CUDA) {
   }
   std::vector<size_t> expect_dims{0, 1, 2, 2, 3, 4, 5, 2, 2, 2, 6};
   EXPECT_EQ(dims, expect_dims);
+
+  auto merged_dim = tv->axis(2)->toString(0);
+  EXPECT_EQ(
+      merged_dim.substr(merged_dim.find("{")),
+      "{( 23 * ( 19 * ( 29 * ( 5 * 7 ) ) ) )}");
```
Comment (jjsjann123) [inline test/test_gpu_utils.cpp:66]: lol is there a better way to check the order of merged IterDomain than string comparison? @zasdfgbnm

File: csrc/scheduler/utils.cpp (hunk 2 above)
Comment (jjsjann123) [reply csrc/scheduler/utils.cpp]: I tried to put more comment, feel free to yell at me if it's not helping :stuck_out_tongue:

**Comment by jjsjann123** [comment]
!build

File: csrc/scheduler/utils.cpp [hunk 4]
```diff
@@ -170,21 +170,50 @@ std::optional<size_t> mergeDims(
...
+  //
+  // Otherwise this could results in misaligned memory access due to indexing.
+  // This is because we compute vectorization width before applying scheduling
+  // transformations.
+  for (size_t i = 1; i < to_merge.size(); i++) {
+    auto outer = to_merge[i];
+    // If outer > inner, the merge order conflicts with their order in leaf
+    // domain
+    if (outer > inner) {
+      // NOTE: reorder here is necessary.
```
Comment (naoyam) [inline csrc/scheduler/utils.cpp:188]: Maybe rephrase like:

```
// NOTE: reorder here is not strictly necessary as `TensorView::merge` automatically does if the first axis position is larger than the second. However we also need to keep track of the actual outer position for the remaining merge operations as well as for return value.
```

Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:188]: I think you mean `swap` isn't necessary. But the `reorder` call is.
The merged IterDomain would have the wrong order otherwise.

File: test/test_gpu_utils.cpp (hunk 3 above)
Comment (naoyam) [reply test/test_gpu_utils.cpp:66]: Maybe you could use `merged_dim->definition()->outer()` and `merged_dim->definition()->inner()`. Or you could use `IterVisitor::getInputsTo()`.

File: csrc/scheduler/utils.cpp (hunk 4 above)
Comment (naoyam) [reply csrc/scheduler/utils.cpp:188]: Hmm, I didn't think about the reordering. Let me think about it more carefully...
